core/src/main/scala/org/http4s/internal/package.scala

```diff
@@ -23,4 +24,78 @@ package object internal {
       case Left(e) => IO(logger.error(e)("Error in asynchronous callback"))
       case Right(_) => IO.unit
     }
+
+  /** Hex encoding digits. Adapted from apache commons Hex.encodeHex **/
+  private val Digits: Array[Char] =
+    Array('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F')
+
+  /** Encode a byte Array into a hexadecimal string
+    *
+    * @param data the array
+    * @return a hexadecimal encoded string
+    */
+  private[http4s] final def encodeHexString(data: Array[Byte]): String =
+    new String(encodeHex(data))
+
+  /** Encode a string to a Hexadecimal string representation
+    * Adapted from apache commons Hex.encodeHex
+    */
+  private[http4s] final def encodeHex(data: Array[Byte]): Array[Char] = {
```

Comment:
I assume this is something people learn when they go to school.

Reply:
I've written this code more than once unfortunately to avoid pulling in a dependency just for this.